### PR TITLE
Fuzzing via honggfuzz

### PR DESCRIPTION
**Pull request by Nereuxofficial**
Originally i was not aware of the crate, so i wanted to implement JPEG-XL parsing in Rust myself, then i discovered this crate and decided to improve the existing crate(which seems very well designed). 

I noticed that it did not have a fuzzing setup and decided to add one and fuzzed it while i was at it.

I discovered two crashes and added tests for them:
1. Integer overflow in a multiplication in compute_default_width
2. Very large memory allocation(probably related to #54 )
Additionally i did not find a security policy inplace so i will just report it via this PR.

I would love to fix the first one, but would like to coordinate as this would probably make `compute_default_width` return a result.

I found out later there is another PR adding Fuzzing support, however development seems to have stalled. However i plan on implementing more fuzzing coverage and a corpus, as seen in that PR and would also be fine with further improving that one, if AFL seems like the better solution. (See #31 )

### CODE REVIEW

File: crates/jxl-oxide/fuzz/hfuzz_workspace/fuzz_decode/HONGGFUZZ.REPORT.TXT [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ <0x000055555556821a> [func:UNKNOWN file: line:0 module:/home/benedikt/RustroverProjects/jxl-oxide/crates/jxl-oxide/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/release/fuzz_decode]
+ <0x00005555555685e8> [func:UNKNOWN file: line:0 module:/home/benedikt/RustroverProjects/jxl-oxide/crates/jxl-oxide/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/release/fuzz_decode]
+ <0x00005555555f05d7> [func:UNKNOWN file: line:0 module:/home/benedikt/RustroverProjects/jxl-oxide/crates/jxl-oxide/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/release/fuzz_decode]
+ <0x00005555555685bc> [func:UNKNOWN file: line:0 module:/home/benedikt/RustroverProjects/jxl-oxide/crates/jxl-oxide/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/release/fuzz_decode]
+ <0x00007ffff7caccd0> [func:UNKNOWN file: line:0 module:/usr/lib/libc.so.6]
+ <0x00007ffff7cacd8a> [func:UNKNOWN file: line:0 module:/usr/lib/libc.so.6]
+ <0x0000555555567b95> [func:UNKNOWN file: line:0 module:/home/benedikt/RustroverProjects/jxl-oxide/crates/jxl-oxide/fuzz/hfuzz_target/x86_64-unknown-linux-gnu/release/fuzz_decode]
+=====================================================================
```
Comment (tirr-c) [inline crates/jxl-oxide/fuzz/hfuzz_workspace/fuzz_decode/HONGGFUZZ.REPORT.TXT:104]: I guess this shouldn't be included. Maybe ignore `hfuzz_workspace` if we can specify custom corpus directory?

File: crates/jxl-oxide/fuzz/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+[package.metadata]
+cargo-fuzz = true
+
+[dependencies]
+honggfuzz = "0.5.55"
+jxl-oxide = {path = ".."}
+
+# Prevent this from interfering with workspaces
+[workspace]
+members = ["."]
```
Comment (tirr-c) [inline crates/jxl-oxide/fuzz/Cargo.toml:16]: How about having the fuzzer as a workspace member? You can move this crate to `crates/jxl-oxide-fuzz` and add it to toplevel `Cargo.toml`.

**Comment by Nereuxofficial** [comment]
Upon further fuzzing to have higher coverage(now 28% compared to 12% previously) i found another issue. The PR will be updated shortly with a test for it. 

There is an out of bounds index access in dequant.rs [here](https://github.com/tirr-c/jxl-oxide/blob/0d041dd1027b899cc7cb33101f9a6b44b94a4789/crates/jxl-vardct/src/dequant.rs#L546C56-L546C61).

**Review by tirr-c — COMMENTED** [review]
Thanks for contributing! Having a fuzzer would be nice.

Regarding `compute_default_width`: the computed `width` should be within `u32` range, as `height <= (1 << 30)` (from `1 + u(30)`, meaning that it reads 30 bits and adds 1) and the aspect ratio is at most 2. Casting `height` to `u64` will do the thing. Maybe add an assertion to `height`?

File: crates/jxl-oxide/fuzz/Cargo.toml (hunk 2 above)
Comment (Nereuxofficial) [reply crates/jxl-oxide/fuzz/Cargo.toml:16]: Sounds like a good idea. Will do that!

File: crates/jxl-oxide/fuzz/hfuzz_workspace/fuzz_decode/HONGGFUZZ.REPORT.TXT (hunk 1 above)
Comment (Nereuxofficial) [reply crates/jxl-oxide/fuzz/hfuzz_workspace/fuzz_decode/HONGGFUZZ.REPORT.TXT:104]: Yes that is possible. The report will be ignored, however the other files with the admittedly random seeming names have to be included unless the test cases are done another way.

File: crates/jxl-oxide/fuzz/fuzz_targets/decode.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    fn test_multiply_with_overflow_height() {
+        //let data = include_bytes!("../hfuzz_workspace/fuzz_decode/SIGABRT.PC.7ffff7d1383c.STACK.191f89ecd1.CODE.-6.ADDR.0.INSTR.mov____%eax,%ebx.fuzz");
+        //let _ = JxlImage::from_reader(std::io::Cursor::new(data));
+    }
+    #[test]
+    fn test_crash_1() {
+        let data = include_bytes!("../hfuzz_workspace/fuzz_decode/SIGABRT.PC.7ffff7d1383c.STACK.c9ec5b426.CODE.-6.ADDR.0.INSTR.mov____%eax,%ebx.fuzz");
+        let _ = JxlImage::from_reader(std::io::Cursor::new(data));
+    }
+}
```
[the remark this answers is not in the document]
Comment (Nereuxofficial) [reply crates/jxl-oxide/fuzz/fuzz_targets/decode.rs]: Oh i was not aware that was a thing. Is it a problem if the crash files are used with include_bytes from the jxl-oxide-fuzz directory or should they be moved, which could clutter the crate?

[the remark this answers is not in the document]
Comment (tirr-c) [reply crates/jxl-oxide/fuzz/fuzz_targets/decode.rs]: I think it's better to have those files in `jxl-oxide`. The fuzzer is for finding problematic inputs. After such inputs are found, they are used to test `jxl-oxide` itself, not the fuzzer.

File: crates/jxl-oxide/fuzz/hfuzz_workspace/fuzz_decode/HONGGFUZZ.REPORT.TXT (hunk 1 above)
Comment (tirr-c) [reply crates/jxl-oxide/fuzz/hfuzz_workspace/fuzz_decode/HONGGFUZZ.REPORT.TXT]: We can use the crash report and input files to create an issue (which can be automated by some scheduled fuzzer run) or write tests. The workspace directory can be `.gitignore`-ed if we do this; I don't see any reason to leave and commit the directory as is.

**Review by tirr-c — COMMENTED** [review]
You might want to move `test_findings.rs` to the parent directory; only files directly put inside of `tests` directory will run as tests. See [the section about integration tests in the Rust Book][book] for details.

[book]: https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests

**Comment by Nereuxofficial** [comment]
I moved the tests to the integration tests, made a macro for testing them, removed the duplicates from the findings and added a Readme to the fuzz dir to guide new developers.

One thing that is still bothering me is that the fuzz_decode function is now duplicate, since it is needed in the tests of another crate and i do not want to change the public interface of jxl-oxide. Any idea how to elegantly solve this?

Additionally i could make a github workflow to fuzz the crate with a configurable timeout, which reports new findings via an issue but i would make a new PR for that, since the changes are fairly independent.

**Comment by tirr-c** [comment]
Uh, another `Cargo.lock` should not happen... Does `cargo hfuzz` create `Cargo.lock` inside the fuzzer directory?

File: crates/jxl-oxide/tests/test_fuzz_findings.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+fn fuzz_decode(data: &[u8]) {
+    if let Ok(mut image) = JxlImage::from_reader(std::io::Cursor::new(data)) {
+        let header = image.image_header();
+        let max_size = u32::max(header.size.width, header.size.height);
+
+        // Skip huge images
+        if max_size > 65536 {
+            return;
+        }
```
Comment (tirr-c) [inline crates/jxl-oxide/tests/test_fuzz_findings.rs:11]: For example, we don't need this part since we're not fuzzing here.

**Review by tirr-c — COMMENTED** [review]
> One thing that is still bothering me is that the fuzz_decode function is now duplicate, since it is needed in the tests of another crate and i do not want to change the public interface of jxl-oxide. Any idea how to elegantly solve this?

I feel like some duplicate code is okay, as sometimes the fuzzer and actual test harness do slightly different things.

File: crates/jxl-oxide/tests/test_fuzz_findings.rs (hunk 4 above)
Comment (Nereuxofficial) [reply crates/jxl-oxide/tests/test_fuzz_findings.rs]: Oh, fixed now. Thanks!

**Comment by Nereuxofficial** [comment]
> Uh, another `Cargo.lock` should not happen... Does `cargo hfuzz` create `Cargo.lock` inside the fuzzer directory?

No, it was from the previous directory and i thought it should be committed but deleting it and recompiling it, it does not create it now

File: crates/jxl-oxide/tests/test_fuzz_findings.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+use jxl_oxide::JxlImage;
+
+fn fuzz_decode(data: &[u8]) {
+    if let Ok(mut image) = JxlImage::from_reader(std::io::Cursor::new(data)) {
+        let header = image.image_header();
+        let max_size = u32::max(header.size.width, header.size.height);
```
Comment (tirr-c) [inline crates/jxl-oxide/tests/test_fuzz_findings.rs:6]: We don't need `max_size` either.

File: crates/jxl-oxide-fuzz/.gitignore [hunk 6]
```diff
@@ -0,0 +1,4 @@
+hfuzz_workspace/*/*
+hfuzz_workspace/HONGGFUZZ.REPORT.TXT
+hfuzz_target
+target
```
Comment (tirr-c) [inline crates/jxl-oxide-fuzz/.gitignore:4]:
```suggestion
hfuzz_workspace
hfuzz_target
```

There are some files committed inside `hfuzz_workspace`. Don't forget to move (or delete) those!

File: crates/jxl-oxide-fuzz/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+name = "jxl-oxide-fuzz"
+version = "0.0.0"
+publish = false
+edition = "2021"
+
+[package.metadata]
+cargo-fuzz = true
+
+[dependencies]
+jxl-oxide = {path = "../jxl-oxide" }
```
Comment (tirr-c) [inline crates/jxl-oxide-fuzz/Cargo.toml:11]:
```suggestion
jxl-oxide = { path = "../jxl-oxide" }
```

**Review by tirr-c — COMMENTED** [review]
Great work! Some nitpicks:

File: crates/jxl-oxide/tests/test_fuzz_findings.rs (hunk 5 above)
Comment (Nereuxofficial) [reply crates/jxl-oxide/tests/test_fuzz_findings.rs:6]: Oh i completely missed that

**Review by tirr-c — APPROVED** [review]
Now we have some remaining work (which can be done in separate PRs):
- Fix crashes. Some crashes are from allocation errors, but there's some chance those errors are from previous integer overflow. Needs investigating.
- Run fuzzer in GitHub Actions workflows.
- Prepare corpus directory for the fuzzer.

Again, thanks for contributing! I'm going to merge this. Feel free to open another PR for additional work.

**Comment by Nereuxofficial** [comment]
> Now we have some remaining work (which can be done in separate PRs):
> - Fix crashes. Some crashes are from allocation errors, but there's some chance those errors are from previous integer overflow. Needs investigating.
> - Run fuzzer in GitHub Actions workflows.
> - Prepare corpus directory for the fuzzer.
> 
> Again, thanks for contributing! I'm going to merge this. Feel free to open another PR for additional work.

Sounds good! Thanks for helping me!

File: crates/jxl-image/src/lib.rs [hunk 8]
```diff
@@ -503,65 +515,289 @@ impl<Ctx> Bundle<Ctx> for BitDepth {
 #[allow(clippy::excessive_precision)]
 impl ImageMetadata {
     const D_UP2: [f32; 15] = [
-        -0.01716200, -0.03452303, -0.04022174, -0.02921014, -0.00624645,
-        0.14111091, 0.28896755, 0.00278718, -0.01610267, 0.56661550,
-        0.03777607, -0.01986694, -0.03144731, -0.01185068, -0.00213539,
+        -0.01716200,
```
Comment (EugeneVert) [inline crates/jxl-image/src/lib.rs:518]: @tirr-c, Shall we use rustfmt for the new code?

Comment (Nereuxofficial) [reply crates/jxl-image/src/lib.rs:518]: oh i did not want to format that file. That blew up the lines. Gonna turn autoformatting off in my IDE

Comment (tirr-c) [reply crates/jxl-image/src/lib.rs:518]: @EugeneVert I'd like to do that, yes. Rustfmt properly handles let-else syntax now, so I guess it's time to do formatting!

**Pull request merged**